### PR TITLE
[autochunk] support transformer

### PR DESCRIPTION
**Pull request by oahzxl**
support transformer for autochunk
1. support some op for transformer, such as assert, type, split and etc.
1. fix some bugs in chunk region search
1. rewrite tests, make them simpler

### CODE REVIEW

File: colossalai/autochunk/search_chunk.py [hunk 1]
```diff
@@ -114,6 +120,12 @@ def _search_max_chunk_region(self, active_node: List, peak_node_idx: int, chunk_
             chunk_region_start (int)
             chunk_region_end (int)
         """
+        # check if peak node already in chunkinfo
+        if chunk_regions is not None:
+            for i in chunk_regions:
+                if i["region"][0] < peak_node_idx <= i["region"][1]:
```
Comment (feifeibear) [inline colossalai/autochunk/search_chunk.py:126]: I guess this line is not what you want
```
i["region"][0] < peak_node_idx and peak_node_idx <= i["region"][1]
```

File: colossalai/autochunk/select_chunk.py [hunk 2]
```diff
@@ -63,33 +61,26 @@ def _select_fit_memory_chunk_region(
         if len(possible_chunk_regions) == 0:
             return None
 
+        max_possible_chunk_region = (min([i["region"][0] for i in possible_chunk_regions]),
+                                     max([i["region"][1] for i in possible_chunk_regions]))
+
         # get mem for chunk region
         regions_dict = []
```
Comment (feifeibear) [inline colossalai/autochunk/select_chunk.py:68]: it is a dict list so the name should XXX_list?

File: colossalai/autochunk/select_chunk.py [hunk 3]
```diff
@@ -63,33 +61,26 @@ def _select_fit_memory_chunk_region(
...
-                        "reorder_chunk_info": cur_region,
-                        "reorder_node_list": cur_node_list,
-                    }
-                )
+                regions_dict.append({
+                    "chunk_info": region,
+                    "chunk_max_mem": cur_chunk_region_max_peak,
+                    "chunk_len": self._get_compute_node_num(region["region"][0], region["region"][1]),
+                    "reorder_chunk_info": cur_region,
+                    "reorder_node_list": cur_node_list,
```
Comment (feifeibear) [inline colossalai/autochunk/select_chunk.py:82]: Can we use a dataclass to replace a dict

Comment (feifeibear) [reply colossalai/autochunk/select_chunk.py:82]: You can update it in the future PR.

File: tests/test_autochunk/test_transformer/test_autochunk_gpt.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
+from functools import partial
+from typing import List, Tuple
+
+import pytest
+import torch
+import torch.multiprocessing as mp
+
+try:
+    from transformers import GPT2Config, GPT2Model
```
Comment (feifeibear) [inline tests/test_autochunk/test_transformer/test_autochunk_gpt.py:9]: We have prepared many prebuilt test models, you do not need to define them by you self.
```
test_models = ['gpt2', 'bert', 'simple_net', 'repeated_computed_layers', 'nested_model', 'albert']

    for model_name in test_models:
        get_components_func = non_distributed_component_funcs.get_callable(model_name)
        model_builder, train_dataloader, _, _, criterion = get_components_func()
 ```

File: tests/test_autochunk/test_alphafold/test_alphafold_utils.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+    if get_chunk_target is not None:
+        chunk_found = [i["region"] for i in chunks]
+        chunk_target = get_chunk_target()[max_memory]
+        assert chunk_found == chunk_target, "found regions %s doesn't equal target regions %s" % (
+            str(chunk_found),
+            str(chunk_target),
+        )
+
+    gpc.destroy()
```
[the remark this answers is not in the document]
Comment (oahzxl) [reply tests/test_autochunk/test_alphafold/test_alphafold_utils.py:124]: sure

File: colossalai/autochunk/select_chunk.py (hunk 2 above)
Comment (oahzxl) [reply colossalai/autochunk/select_chunk.py:68]: it is a list[dict], I will update its name

File: colossalai/autochunk/select_chunk.py (hunk 3 above)
Comment (oahzxl) [reply colossalai/autochunk/select_chunk.py:82]: good suggestion, I will update in future

**Review by feifeibear — APPROVED** [review]

**Pull request merged**